numpy/core/src/_simd/_simd.c

```diff
@@ -0,0 +1,73 @@
+#include "_simd.h"
+
+PyMODINIT_FUNC PyInit__simd(void)
+{
+    static struct PyModuleDef defs = {
+        .m_base = PyModuleDef_HEAD_INIT,
+        .m_name = "_simd",
+        .m_size = -1
+    };
+    if (npy_cpu_init() < 0) {
+        return NULL;
+    }
+    PyObject *m = PyModule_Create(&defs);
+    if (m == NULL) {
+        return NULL;
+    }
+    PyObject *targets = PyDict_New();
+    if (targets == NULL) {
+        goto err;
+    }
+    if (PyModule_AddObject(m, "targets", targets) < 0) {
+        Py_DECREF(targets);
+        goto err;
+    }
+    // add keys for non-supported optimizations with None value
+    #define ATTACH_MODULE(TESTED_FEATURES, TARGET_NAME, MAKE_MSVC_HAPPY)       \
+        {                                                                      \
+            PyObject *simd_mod;                                                \
+            if (!TESTED_FEATURES) {                                            \
+                Py_INCREF(Py_None);                                            \
+                simd_mod = Py_None;                                            \
+            } else {                                                           \
+                simd_mod = NPY_CAT(simd_create_module_, TARGET_NAME)();        \
+                if (simd_mod == NULL) {                                        \
+                    goto err;                                                  \
+                }                                                              \
+            }                                                                  \
+            const char *target_name = NPY_TOSTRING(TARGET_NAME);               \
+            if (PyDict_SetItemString(targets, target_name, simd_mod) < 0) {    \
```

Comment:
Tempting to make `targets` just a list, which can be done with:
```suggestion
    PyObject *targets = PyList_New(0);
    if (targets == NULL) {
        goto err;
    }
    if (PyModule_AddObject(m, "targets", targets) < 0) {
        Py_DECREF(targets);
        goto err;
    }
    // add keys for non-supported optimizations with None value
    #define ATTACH_MODULE(TESTED_FEATURES, TARGET_NAME, MAKE_MSVC_HAPPY)       \
        {                                                                      \
            PyObject *simd_mod;                                                \
            if (!TESTED_FEATURES) {                                            \
                Py_INCREF(Py_None);                                            \
                simd_mod = Py_None;                                            \
            } else {                                                           \
                simd_mod = NPY_CAT(simd_create_module_, TARGET_NAME)();        \
                if (simd_mod == NULL) {                                        \
                    goto err;                                                  \
                }                                                              \
            }                                                                  \
            if (PyList_Append(targets, simd_mod) < 0) {    \
```
Then you can just us `targets[i].__name__`, or maybe add your own `targets[i].pretty_name` if `__name__` is too ugly for you.

Reply:
I guess there's a miss understanding here, why do we need a dictionary instead of a list?

1- we need to verify the enabled targets through infrastructure, `--simd-test`
2- we need to verify the CPU support.

for example:
```Python3

from numpy.core._simd import targets
for target_name, npyv in targets.items(): # get all enabled SIMD targets
    if npyv is None:
        # a module is compiled for this target but the running CPU doesn't support it
        print(f"your CPU doesn't support {target_name}, try Intel SDE or QEMU")
        continue
    if not npyv.simd: # equivlent to C macro "NPY_SIMD"
        # now we got the module and CPU support the target but our NPYV don't
        print("why did you compile it via --simd-test?")
        continue
    a = npyv.load_f32(range(npyv.nlanes_f32))
    print(f"load a f32 vector for target {target_name}", a)

````

Reply:
Sure, but you could also write:
```
from numpy.core._simd import targets
for npyv in targets: # get all enabled SIMD targets
    if npyv is None:
        # a module is compiled for this target but the running CPU doesn't support it
        print(f"your CPU doesn't support {target_name}, try Intel SDE or QEMU")
        continue
    if not npyv.simd: # equivlent to C macro "NPY_SIMD"
        # now we got the module and CPU support the target but our NPYV don't
        print("why did you compile it via --simd-test?")
        continue
    a = npyv.load_f32(range(npyv.nlanes_f32))
    print(f"load a f32 vector for target {npyv.__name__}", a)  # or `npyv.pretty_name`, etc
```

Reply:
How can I get the target name for None value? check the first case in the loop

Reply:
Thank you, that was what I missed - your point is that you want to explicitly loop over the targets that are absent too. In that case, a dict is probably sensible after all.